cmd/snap/complete.go

```diff
@@ -36,6 +36,84 @@ import (
 	"github.com/snapcore/snapd/snap"
 )
 
+// SnapAndApp holds a snap name and an application name
+type SnapAndApp struct {
+	Snap   string
+	App    string
+	hasDot bool
+}
+
+func (sa *SnapAndApp) FullName() string {
+	if sa.hasDot {
+		return fmt.Sprint(sa.Snap, ".", sa.App)
+	} else {
+		return sa.Snap
+	}
+}
+
+// UnmarshalFlag unmarshals the snap and application name. The following
+// combinations are allowed:
+// * <snap>.<app>
+// * <snap>
+// Every other combination results in an error.
+func (sa *SnapAndApp) UnmarshalFlag(value string) error {
+	parts := strings.Split(value, ".")
+	sa.Snap = ""
+	sa.App = ""
+	sa.hasDot = false
+	switch len(parts) {
+	case 1:
+		sa.Snap = parts[0]
+	case 2:
+		sa.Snap = parts[0]
+		sa.App = parts[1]
+		sa.hasDot = true // allows to know if it is "snap." or "snap.XXXX"
+	}
+	if sa.Snap == "" && sa.App != "" {
+		return fmt.Errorf(i18n.G("invalid value: %q (want snap.name or snap)"), value)
+	}
+	return nil
+}
+
+func (s SnapAndApp) Complete(match string) []flags.Completion {
+	var matchSnap SnapAndApp
+	matchSnap.UnmarshalFlag(match)
+	if !matchSnap.hasDot {
+		// No dot in match, so complete with snap names
+		installedSnaps, err := mkClient().List(nil, nil)
+		if err != nil {
+			return nil
+		}
+		ret := make([]flags.Completion, 0, len(installedSnaps))
```

Comment:
this could perhaps be too optimistic. Not all of the snaps can possibly match, maybe this could be tweaked to something like:
```suggestion
		// we're expecting only a subset of all snaps to match, so start
		// with a reasonable guess for capacity and allocate more as we
		// go through the list and find more matches
		const matchesCapacityGuess = 5
		ret := make([]flags.Completion, 0, matchesCapacityGuess)
```

I guess it doesn't matter that much since we already allocated the memory to hold a list of all installed snaps, but let's not allocate more if there isn't a clear need.

Reply:
Changed to no reservation.

Reply:
It's not needed, because it's called only "sometimes", so the speed gain is negligible.